File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/MusicController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.MusicFacade;
import cz.vhromada.catalog.facade.to.MusicTO;
import cz.vhromada.validators.exceptions.RecordNotFoundException;
import cz.vhromada.validators.exceptions.ValidationException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for music.
 *
 * @author Vladimir Hromada
 */
@Controller("musicController")
@RequestMapping("/music")
public class MusicController extends JsonController {

	@Autowired
	@Qualifier("musicFacade")
	private MusicFacade musicFacade;

	/** Creates new data. */
	@RequestMapping(value = "/new", method = RequestMethod.GET)
	@ResponseBody
	public void newData() {
		musicFacade.newData();
	}

	/**
	 * Returns list of music.
	 *
	 * @return list of music
	 */
	@RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
	@ResponseBody
	public String getMusic() {
		return serialize(musicFacade.getMusic());
	}

	/**
	 * Returns music with ID or null if there isn't such music.
	 *
	 * @param id ID
	 * @return music with ID or null if there isn't such music
	 * @throws IllegalArgumentException if ID is null
	 */
	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
	@ResponseBody
	public String getMusic(@PathVariable("id") final Integer id) {
		return serialize(musicFacade.getMusic(id));
	}

	/**
	 * Adds music. Sets new ID and position.
	 *
	 * @param music music
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID isn't null
	 *                                  or name is null
	 *                                  or name is empty string
	 *                                  or URL to english Wikipedia page about music is null
	 *                                  or URL to czech Wikipedia page about music is null
	 *                                  or count of media isn't positive number
	 *                                  or count of songs is negative number
	 *                                  or total length of songs is null
	 *                                  or total length of songs is negative number
	 *                                  or note is null
	 */
	@RequestMapping(value = "/add", method = RequestMethod.POST)
	@ResponseBody
	public void add(final String music) {
		musicFacade.add(deserialize(music, MusicTO.class));
	}

	/**
	 * Updates music.
	 *
	 * @param music new value of music
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID is null
	 *                                  or name is null
	 *                                  or name is empty string
	 *                                  or URL to english Wikipedia page about music is null
	 *                                  or URL to czech Wikipedia page about music is null
	 *                                  or count of media isn't positive number
	 *                                  or count of songs is negative number
	 *                                  or total length of songs is null
	 *                                  or total length of songs is negative number
	 *                                  or note is null
	 * @throws RecordNotFoundException  if music doesn't exist in data storage
	 */
	@RequestMapping(value = "/update", method = RequestMethod.POST)
	@ResponseBody
	public void update(final String music) {
		musicFacade.update(deserialize(music, MusicTO.class));
	}

	/**
	 * Removes music.
	 *
	 * @param music music
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if music doesn't exist in data storage
	 */
	@RequestMapping(value = "/remove", method = RequestMethod.POST)
	@ResponseBody
	public void remove(final String music) {
		musicFacade.remove(deserialize(music, MusicTO.class));
	}

	/**
	 * Duplicates music.
	 *
	 * @param music music
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if music doesn't exist in data storage
	 */
	@RequestMapping(value = "/duplicate", method = RequestMethod.POST)
	@ResponseBody
	public void duplicate(final String music) {
		musicFacade.duplicate(deserialize(music, MusicTO.class));
	}

	/**
	 * Moves music in list one position up.
	 *
	 * @param music music
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID is null
	 *                                  or music can't be moved up
	 * @throws RecordNotFoundException  if music doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveUp", method = RequestMethod.POST)
	@ResponseBody
	public void moveUp(final String music) {
		musicFacade.moveUp(deserialize(music, MusicTO.class));
	}

	/**
	 * Moves music in list one position down.
	 *
	 * @param music music
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID is null
	 *                                  or music can't be moved down
	 * @throws RecordNotFoundException  if music doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveDown", method = RequestMethod.POST)
	@ResponseBody
	public void moveDown(final String music) {
		musicFacade.moveDown(deserialize(music, MusicTO.class));
	}

	/**
	 * Returns true if music exists.
	 *
	 * @param music music
	 * @return true if music exists
	 * @throws IllegalArgumentException if music is null
	 * @throws ValidationException      if ID is null
	 */
	@RequestMapping(value = "/exists", method = RequestMethod.POST)
	@ResponseBody
	public String exists(final String music) {
		return serialize(musicFacade.exists(deserialize(music, MusicTO.class)));
	}

	/** Updates positions. */
	@RequestMapping(value = "/updatePositions", method = RequestMethod.GET)
	@ResponseBody
	public void updatePositions() {
		musicFacade.updatePositions();
	}

	/**
	 * Returns total count of media.
	 *
	 * @return total count of media
	 */
	@RequestMapping(value = "/totalMedia", method = RequestMethod.GET)
	@ResponseBody
	public String getTotalMediaCount() {
		return serialize(musicFacade.getTotalMediaCount());
	}

	/**
	 * Returns total length of all music.
	 *
	 * @return total length of all music
	 */
	@RequestMapping(value = "/totalLength", method = RequestMethod.GET)
	@ResponseBody
	public String getTotalLength() {
		return serialize(musicFacade.getTotalLength());
	}

	/**
	 * Returns count of songs from all music.
	 *
	 * @return count of songs from all music
	 */
	@RequestMapping(value = "/songsCount", method = RequestMethod.GET)
	@ResponseBody
	public String getSongsCount() {
		return serialize(musicFacade.getSongsCount());
	}

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/MusicController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Music
import com.github.vhromada.catalog.facade.MusicFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for music.
 *
 * @author Vladimir Hromada
 */
@RestController("musicController")
@RequestMapping("/catalog/music")
class MusicController(
    private val musicFacade: MusicFacade
) : AbstractController() {

    /**
     * Returns list of music.
     *
     * @return list of music
     */
    @GetMapping
    fun getMusic(): List<Music> {
        return processResult(result = musicFacade.getAll())!!
    }

    /**
     * Creates new data.
     */
    @PostMapping("/new")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun newData() {
        processResult(result = musicFacade.newData())
    }

    /**
     * Returns music with ID.
     *
     * @param id ID
     * @return music with ID
     */
    @GetMapping("/{id}")
    fun getMusic(@PathVariable("id") id: Int): Music {
        return processResult(result = musicFacade.get(id = id))!!
    }

    /**
     * Adds music. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Name is null
     *  * Name is empty string
     *  * URL to english Wikipedia page about music is null
     *  * URL to czech Wikipedia page about music is null
     *  * Count of media isn't positive number
     *  * Other data is null
     *  * Note is null
     *
     * @param music music
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@RequestBody music: Music) {
        processResult(result = musicFacade.add(data = music))
    }

    /**
     * Updates music.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * URL to english Wikipedia page about music is null
     *  * URL to czech Wikipedia page about music is null
     *  * Count of media isn't positive number
     *  * Other data is null
     *  * Note is null
     *  * Music doesn't exist in data storage
     *
     * @param music new value of music
     */
    @PostMapping("/update")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun update(@RequestBody music: Music) {
        processResult(result = musicFacade.update(data = music))
    }

    /**
     * Removes music.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *
     * @param id ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("id") id: Int) {
        processResult(result = musicFacade.remove(id = id))
    }

    /**
     * Duplicates music.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@PathVariable("id") id: Int) {
        processResult(result = musicFacade.duplicate(id = id))
    }

    /**
     * Moves music in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Music can't be moved up
     *  * Music doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@PathVariable("id") id: Int) {
        processResult(result = musicFacade.moveUp(id = id))
    }

    /**
     * Moves music in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Music can't be moved down
     *  * Music doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@PathVariable("id") id: Int) {
        processResult(result = musicFacade.moveDown(id = id))
    }

    /**
     * Updates positions.
     */
    @PostMapping("/updatePositions")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updatePositions() {
        processResult(result = musicFacade.updatePositions())
    }

    /**
     * Returns total count of media.
     *
     * @return total count of media
     */
    @GetMapping("/totalMedia")
    fun getTotalMediaCount(): Int {
        return processResult(result = musicFacade.getTotalMediaCount())!!
    }

    /**
     * Returns total length of all music.
     *
     * @return total length of all music
     */
    @GetMapping("/totalLength")
    fun getTotalLength(): Int {
        return processResult(result = musicFacade.getTotalLength())!!.length
    }

    /**
     * Returns count of songs from all music.
     *
     * @return count of songs from all music
     */
    @GetMapping("/songsCount")
    fun geSongsCount(): Int {
        return processResult(result = musicFacade.getSongsCount())!!
    }

}
